Fixed #24170 -- Implemented decompress for BaseRangeField widgets

ngzhian opened the pull request:
Introduce a new range widget `RangeWidget` to be used for BaseRangeField as part of new Postgres ranges supported.

Suggested by @charettes [here](https://code.djangoproject.com/ticket/24170#comment:3)

collinanderson commented on django/contrib/postgres/widgets.py:
```diff
@@ -0,0 +1,13 @@
+from django.forms.widgets import MultiWidget
+from django.forms.fields import DateTimeInput
```
Unused import?

ngzhian replied:
@collinanderson thanks for that, fixed and rebased

MarkusH commented:
Could you add a note about that widget in the 1.8 release notes (`docs/releases/1.8.txt`) and in the reference documentation (`docs/ref/contrib/postgres`).

PR is lacking tests, otherwise, looks good.

ngzhian commented:
@MarkusH I added a note in reference documentation, I hope it reads okay. I'm not sure how we should make a note of this widget in the 1.8 release notes, because it's just a short paragraph directing users to read the reference docs -- so I don't think mentioning it is necessary.
Added some test in my latest commit, I just initialized a `DateTimeRangeField` and asserted the result of rendering its widget.

charettes commented:
> Could you add a note about that widget in the 1.8 release notes.

I also don't think it should be necessary since this is mainly an implementation detail of a bugfix.

charettes commented:
Buildbot, test this please.

MarkusH commented:
Well, it is only a short documentation note, but I'd expect developers to find this widget in the docs if they want to override some functionality.

ngzhian commented:
@MarkusH I added a link to Widget to the 1.8 release doc in the section about Postgres specific features.

timgraham commented on docs/releases/1.8.txt:
```diff
@@ -73,7 +73,8 @@ New PostgreSQL specific functionality
 as :class:`~django.contrib.postgres.fields.ArrayField`,
-:class:`~django.contrib.postgres.fields.HStoreField`, :ref:`range-fields`, and
+:class:`~django.contrib.postgres.fields.HStoreField`, :ref:`range-fields`,
+:class:`~django.contrib.postgres.widgets.RangeWidget`, and
```
This is out of place -- form fields aren't even mentioned in the release notes and widgets are even more "obscure". I agree with Simon that a note isn't needed here.

timgraham commented on docs/ref/contrib/postgres/forms.txt:
```diff
@@ -199,3 +199,28 @@ DateRangeField
+.. currentmodule:: django.contrib.postgres.widgets
+
+Widgets
+------------
```
underlines should match heading length

timgraham commented on django/contrib/postgres/widgets.py:
```diff
@@ -0,0 +1,12 @@
+from django.forms.widgets import MultiWidget
+
+
+class RangeWidget(MultiWidget):
```
This should likely live in `forms.ranges`

timgraham commented on docs/ref/contrib/postgres/forms.txt:
```diff
@@ -199,3 +199,28 @@ DateRangeField
+
+.. class:: RangeWidget(base_widget, attrs=None)
+
+    Widget used by all of the Range Fields as mentioned above.
```
"range fields" with no caps
chop "as mentioned above"
Add "All of these fields use RangeWidget." above in the description of "Ranges Fields" after "not greater than the upper bound."

timgraham commented on tests/postgres_tests/test_ranges.py:
```diff
@@ -374,3 +374,12 @@ def test_model_field_formfield_datetime(self):
+class TestWidget(TestCase):
+    def test_range_widget(self):
+        f = pg_forms.ranges.DateTimeRangeField()
+        self.assertHTMLEqual(f.widget.render('datetimerange', ''), '<input type="text" name="datetimerange_0" /><input type="text" name="datetimerange_1" />')
```
Please wrap long lines for easier review:
```
self.assertHTMLEqual(
    f.widget.render('datetimerange', ''),
    '<input ...'
)
```

ngzhian commented:
@timgraham thanks for the review, not sure if you would like me to squash all these commits and rebase onto master?

MarkusH commented on docs/ref/contrib/postgres/forms.txt:
```diff
@@ -199,3 +200,26 @@ DateRangeField
     :class:`~django.contrib.postgres.fields.DateRangeField`.
+
+Widgets
+------------
```
The underlines should match the heading length.

MarkusH commented:
LGTM

charettes commented:
Could you squash and rebase your commits on top of master. Also, use past tense in the commit messages (Implement -> Implemented).

ngzhian commented:
@charettes done, edited the pull request title for consistency as well, thanks!

MarkusH commented:
Buildbot, test this please.

charettes commented:
@MarkusH, oops merged before I saw you wanted to run CI on it again.

MarkusH commented:
@charettes No problem. I'm still a bit cautious before merging PRs ;)
